# Phishing URL Detector

A machine learning project that classifies URLs as phishing or legitimate, built on the [PhiUSIIL Phishing URL Dataset](https://www.kaggle.com/datasets/sharmageetika/phishing-url-dataset) (235,795 URLs). Includes custom URL-text feature engineering, model comparison, error analysis, and a live Flask web app (`app.py`).

In [1]:
import pandas as pd
df = pd.read_csv('PhiUSIIL_Phishing_URL_Dataset.csv')
df.head()

,FILENAME,URL,URLLength,Domain,DomainLength,IsDomainIP,TLD,URLSimilarityIndex,CharContinuationRate,TLDLegitimateProb,...,Pay,Crypto,HasCopyrightInfo,NoOfImage,NoOfCSS,NoOfJS,NoOfSelfRef,NoOfEmptyRef,NoOfExternalRef,label
0,521848.txt,https://www.southbankmosaics.com,31,www.southbankmosaics.com,24,0,com,100.0,1.000000,0.522907,...,0,0,1,34,20,28,119,0,124,1
1,31372.txt,https://www.uni-mainz.de,23,www.uni-mainz.de,16,0,de,100.0,0.666667,0.032650,...,0,0,1,50,9,8,39,0,217,1
2,597387.txt,https://www.voicefmradio.co.uk,29,www.voicefmradio.co.uk,22,0,uk,100.0,0.866667,0.028555,...,0,0,1,10,2,7,42,2,5,1
3,554095.txt,https://www.sfnmjournal.com,26,www.sfnmjournal.com,19,0,com,100.0,1.000000,0.522907,...,1,1,1,3,27,15,22,1,31,1
4,151578.txt,https://www.rewildingargentina.org,33,www.rewildingargentina.org,26,0,org,100.0,1.000000,0.079963,...,1,0,1,244,15,34,72,1,85,1


In [2]:
print(df.shape)
df.info()

(235795, 56)
<class 'pandas.DataFrame'>
RangeIndex: 235795 entries, 0 to 235794
Data columns (total 56 columns):
 #   Column                      Non-Null Count   Dtype  
---  ------                      --------------   -----  
 0   FILENAME                    235795 non-null  str    
 1   URL                         235795 non-null  str    
 2   URLLength                   235795 non-null  int64  
 3   Domain                      235795 non-null  str    
 4   DomainLength                235795 non-null  int64  
 5   IsDomainIP                  235795 non-null  int64  
 6   TLD                         235795 non-null  str    
 7   URLSimilarityIndex          235795 non-null  float64
 8   CharContinuationRate        235795 non-null  float64
 9   TLDLegitimateProb           235795 non-null  float64
 10  URLCharProb                 235795 non-null  float64
 11  TLDLength                   235795 non-null  int64  
 12  NoOfSubDomain               235795 non-null  int64  
 13  HasObfuscati

## Exploring the Data

In [3]:
print(df['label'].value_counts())
print(df['label'].value_counts(normalize=True))

label
1    134850
0    100945
Name: count, dtype: int64
label
1    0.571895
0    0.428105
Name: proportion, dtype: float64


In [4]:
pd.set_option('display.max_colwidth', 100)
print("LEGITIMATE examples:")
print(df[df['label'] == 1][['URL']].sample(5, random_state=1))
print()
print("PHISHING examples:")
print(df[df['label'] == 0][['URL']].sample(5, random_state=1))

LEGITIMATE examples:
                                      URL
93003             https://www.levelup.com
194803        https://www.notjusttoyz.com
117359  https://www.discover-suriname.com
92456       https://www.ecolabelindex.com
160985        https://www.cinematheque.fr

PHISHING examples:
                                                                                                        URL
220175  https://tracking.selfserviceib.com/tracking/1/click/ejorccw7cfkwhilpbcpgu7mn1hs1pqyi7ko2fgolwhjz...
25308                                                                            https://maria.susypro.com/
138783  https://exchangeportal-owa-auth-login-aspx-hyttgygh.infura-ipfs.io/ipfs/qmv5almanqswnizvrextgpbq...
90377                                                    https://caseid1008934573489573489.firebaseapp.com/
226844         https://cloudflare-ipfs.com/ipfs/bafybeidmnt5savzueeriznoifcppedx7gf7mobvbtv7mvtijirxwexah6i


In [5]:
print(df.groupby('label')['URLLength'].describe())

          count       mean        std   min   25%   50%   75%     max
label                                                                
0      100945.0  45.720293  61.145523  13.0  26.0  34.0  48.0  6097.0
1      134850.0  26.228610   4.815612  15.0  23.0  26.0  29.0    57.0


In [6]:
print(df.groupby('label')['NoOfSubDomain'].describe())
print()
print(df.groupby('label')['IsHTTPS'].mean())

          count      mean       std  min  25%  50%  75%   max
label                                                        
0      100945.0  1.168894  0.790880  0.0  1.0  1.0  1.0  10.0
1      134850.0  1.161661  0.404076  1.0  1.0  1.0  1.0   4.0

label
0    0.492238
1    1.000000
Name: IsHTTPS, dtype: float64


## Engineering Features From the Raw URL

Built 7 features directly from URL text (not just relying on the dataset's pre-made columns) so I could explain exactly why each one matters, not just that it's in the data.

In [7]:
df['url_length'] = df['URL'].str.len()
df['num_dots'] = df['URL'].str.count(r'\.')
df['num_hyphens'] = df['URL'].str.count('-')
df['num_digits'] = df['URL'].apply(lambda x: sum(c.isdigit() for c in x))
df['num_special_chars'] = df['URL'].str.count(r'[@%&=?]')
df['has_https'] = df['URL'].str.startswith('https://').astype(int)
df['has_ip'] = df['URL'].str.contains(r'\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}').astype(int)

df[['URL', 'url_length', 'num_dots', 'num_hyphens', 'num_digits', 'num_special_chars', 'has_https', 'has_ip']].head()

,URL,url_length,num_dots,num_hyphens,num_digits,num_special_chars,has_https,has_ip
0,https://www.southbankmosaics.com,32,2,0,0,0,1,0
1,https://www.uni-mainz.de,24,2,1,0,0,1,0
2,https://www.voicefmradio.co.uk,30,3,0,0,0,1,0
3,https://www.sfnmjournal.com,27,2,0,0,0,1,0
4,https://www.rewildingargentina.org,34,2,0,0,0,1,0


In [8]:
custom_features = ['url_length', 'num_dots', 'num_hyphens', 'num_digits', 'num_special_chars', 'has_https', 'has_ip']
for col in custom_features:
    print(df.groupby('label')[col].mean())
    print()

label
0    46.238774
1    27.228610
Name: url_length, dtype: float64

label
0    2.384695
1    2.161661
Name: num_dots, dtype: float64

label
0    0.706989
1    0.083204
Name: num_hyphens, dtype: float64

label
0    4.338273
1    0.050597
Name: num_digits, dtype: float64

label
0    0.376661
1    0.000000
Name: num_special_chars, dtype: float64

label
0    0.487354
1    1.000000
Name: has_https, dtype: float64

label
0    0.007014
1    0.000000
Name: has_ip, dtype: float64



## Baseline Models

Trained three models on the same 7 features to compare performance before choosing one to build on.

In [9]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, confusion_matrix

feature_cols = ['url_length', 'num_dots', 'num_hyphens', 'num_digits', 
                 'num_special_chars', 'has_https', 'has_ip']

X = df[feature_cols]
y = df['label']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

model = LogisticRegression(max_iter=1000)
model.fit(X_train, y_train)

y_pred = model.predict(X_test)
print(classification_report(y_test, y_pred))
print(confusion_matrix(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.97      0.89      0.93     20124
           1       0.92      0.98      0.95     27035

    accuracy                           0.94     47159
   macro avg       0.95      0.94      0.94     47159
weighted avg       0.94      0.94      0.94     47159

[[17942  2182]
 [  546 26489]]


In [10]:
from sklearn.ensemble import RandomForestClassifier

rf_model = RandomForestClassifier(n_estimators=100, random_state=42)
rf_model.fit(X_train, y_train)

y_pred_rf = rf_model.predict(X_test)
print(classification_report(y_test, y_pred_rf))
print(confusion_matrix(y_test, y_pred_rf))

              precision    recall  f1-score   support

           0       0.98      0.93      0.95     20124
           1       0.95      0.99      0.97     27035

    accuracy                           0.96     47159
   macro avg       0.97      0.96      0.96     47159
weighted avg       0.96      0.96      0.96     47159

[[18732  1392]
 [  381 26654]]


In [11]:
from xgboost import XGBClassifier

xgb_model = XGBClassifier(n_estimators=100, random_state=42, eval_metric='logloss')
xgb_model.fit(X_train, y_train)

y_pred_xgb = xgb_model.predict(X_test)
print(classification_report(y_test, y_pred_xgb))
print(confusion_matrix(y_test, y_pred_xgb))

              precision    recall  f1-score   support

           0       0.98      0.93      0.95     20124
           1       0.95      0.99      0.97     27035

    accuracy                           0.96     47159
   macro avg       0.97      0.96      0.96     47159
weighted avg       0.96      0.96      0.96     47159

[[18715  1409]
 [  376 26659]]


**Result:** Random Forest and XGBoost both clearly beat Logistic Regression (94% → 96% accuracy), and were close to each other. Chose Random Forest to build on — slightly better phishing recall, and simpler to explain than XGBoost's sequential boosting.

## Error Analysis: What Does the Model Miss?

In [12]:
# Get predictions from the Random Forest model
X_test_with_url = df.loc[X_test.index, ['URL']].copy()
X_test_with_url['actual'] = y_test
X_test_with_url['predicted'] = y_pred_rf

# Find phishing URLs the model missed (actual=0/phishing, predicted=1/legitimate)
missed_phishing = X_test_with_url[(X_test_with_url['actual'] == 0) & (X_test_with_url['predicted'] == 1)]

print(f"Total missed phishing URLs: {len(missed_phishing)}")
print()
pd.set_option('display.max_colwidth', 100)
print(missed_phishing[['URL']].sample(10, random_state=1))

Total missed phishing URLs: 1392

                                              URL
75658             https://pri896users.boxmode.io/
89753            https://logon-review-uk.web.app/
27471               https://porjectf1333.web.app/
136261        https://csvhwubije.firebaseapp.com/
185755     https://www.amzanao.co.ip.prewerd.one/
64243           https://dabrefololokiffe.web.app/
211773  https://metamaskwalletrestore.vercel.app/
43767            https://sbu-tyx.firebaseapp.com/
40195                     https://amz.wpymluy.cn/
203994       https://amazon-vuifadbndf.dnsrd.com/


Looking at the missed phishing URLs, a clear pattern emerged:

- Most were hosted on legitimate platforms (`.web.app`, `.vercel.app`, `.firebaseapp.com`). Since these platforms force HTTPS, my `has_https` feature — normally my best signal — gets fooled.
- Many used brand impersonation through misspelled names (e.g. `amzanao` for Amazon, `metamaskwalletrestore` for MetaMask) instead of digits or odd symbols, which my other features don't catch.

**Fix idea:** add a feature comparing each domain to known brand names using string similarity, to catch typosquatting even when nothing else about the URL looks unusual.

## Fixing the Weakness: Brand Similarity

In [13]:
from difflib import SequenceMatcher

# Common brands phishers impersonate - based on what I saw in my error analysis
common_brands = ['amazon', 'paypal', 'metamask', 'microsoft', 'apple', 'google', 
                  'facebook', 'netflix', 'bank', 'chase', 'wellsfargo', 'coinbase',
                  'binance', 'instagram', 'linkedin', 'ebay', 'walmart']

def brand_similarity(domain):
    domain_clean = domain.lower().replace('www.', '').split('.')[0]
    best_score = 0
    for brand in common_brands:
        score = SequenceMatcher(None, domain_clean, brand).ratio()
        best_score = max(best_score, score)
    return best_score

# Test it on a few examples first
test_domains = ['amzanao', 'amazon', 'metamaskwalletrestore', 'southbankmosaics']
for d in test_domains:
    print(d, '->', brand_similarity(d))

amzanao -> 0.6153846153846154
amazon -> 1.0
metamaskwalletrestore -> 0.5517241379310345
southbankmosaics -> 0.4


In [14]:
df['brand_similarity'] = df['Domain'].apply(brand_similarity)

print(df.groupby('label')['brand_similarity'].mean())
print()
print(df.groupby('label')['brand_similarity'].describe())

label
0    0.355413
1    0.406003
Name: brand_similarity, dtype: float64

          count      mean       std  min       25%       50%       75%  max
label                                                                      
0      100945.0  0.355413  0.123328  0.0  0.285714  0.352941  0.428571  1.0
1      134850.0  0.406003  0.081831  0.0  0.352941  0.400000  0.454545  1.0


In [15]:
feature_cols_v2 = ['url_length', 'num_dots', 'num_hyphens', 'num_digits', 
                    'num_special_chars', 'has_https', 'has_ip', 'brand_similarity']

X2 = df[feature_cols_v2]
y2 = df['label']

X_train2, X_test2, y_train2, y_test2 = train_test_split(X2, y2, test_size=0.2, random_state=42)

rf_model_v2 = RandomForestClassifier(n_estimators=100, random_state=42)
rf_model_v2.fit(X_train2, y_train2)

y_pred_rf_v2 = rf_model_v2.predict(X_test2)
print(classification_report(y_test2, y_pred_rf_v2))
print(confusion_matrix(y_test2, y_pred_rf_v2))

              precision    recall  f1-score   support

           0       0.98      0.95      0.97     20124
           1       0.97      0.99      0.98     27035

    accuracy                           0.97     47159
   macro avg       0.97      0.97      0.97     47159
weighted avg       0.97      0.97      0.97     47159

[[19183   941]
 [  330 26705]]


In [16]:
# Check specifically for the domains that fooled the original model
X_test2_with_url = df.loc[X_test2.index, ['URL']].copy()
X_test2_with_url['actual'] = y_test2
X_test2_with_url['predicted'] = y_pred_rf_v2

typosquat_examples = ['amzanao', 'metamaskwalletrestore', 'porjectf1333', 'sbu-tyx']

for name in typosquat_examples:
    match = X_test2_with_url[X_test2_with_url['URL'].str.contains(name, case=False, na=False)]
    if len(match) > 0:
        print(match)
        print()
    else:
        print(f"{name}: not in test set\n")

                                           URL  actual  predicted
112524  https://www.amzanao.co.ip.pwerous.one/       0          0
185755  https://www.amzanao.co.ip.prewerd.one/       0          0

                                              URL  actual  predicted
211773  https://metamaskwalletrestore.vercel.app/       0          0

                                 URL  actual  predicted
27471  https://porjectf1333.web.app/       0          0

                                    URL  actual  predicted
43767  https://sbu-tyx.firebaseapp.com/       0          1



**Result:** missed phishing dropped from 1,392 → 941 (33% fewer), and I confirmed the exact previously-missed URLs (amzanao, metamaskwalletrestore, porjectf1333) are now caught. One (`sbu-tyx.firebaseapp.com`) still isn't — it doesn't closely resemble any brand in my list, a reasonable limitation given the feature only helps against *known* brand impersonation.

## Pushing Further: Page-Content Features

In [17]:
extra_cols = ['HasPasswordField', 'NoOfExternalRef', 'IsResponsive', 'HasSocialNet', 'LineOfCode', 'NoOfiFrame']
for col in extra_cols:
    print(df.groupby('label')[col].mean())
    print()

label
0    0.053871
1    0.138487
Name: HasPasswordField, dtype: float64

label
0     1.128119
1    85.294601
Name: NoOfExternalRef, dtype: float64

label
0    0.317460
1    0.854364
Name: IsResponsive, dtype: float64

label
0    0.005062
1    0.794557
Name: HasSocialNet, dtype: float64

label
0      65.730467
1    1947.491680
Name: LineOfCode, dtype: float64

label
0    0.084581
1    2.714535
Name: NoOfiFrame, dtype: float64



In [18]:
feature_cols_v3 = ['url_length', 'num_dots', 'num_hyphens', 'num_digits', 
                    'num_special_chars', 'has_https', 'has_ip', 'brand_similarity',
                    'NoOfExternalRef', 'LineOfCode', 'HasSocialNet', 'IsResponsive', 'NoOfiFrame']

X3 = df[feature_cols_v3]
y3 = df['label']

X_train3, X_test3, y_train3, y_test3 = train_test_split(X3, y3, test_size=0.2, random_state=42)

rf_model_v3 = RandomForestClassifier(n_estimators=100, random_state=42)
rf_model_v3.fit(X_train3, y_train3)

y_pred_rf_v3 = rf_model_v3.predict(X_test3)
print(classification_report(y_test3, y_pred_rf_v3))
print(confusion_matrix(y_test3, y_pred_rf_v3))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00     20124
           1       1.00      1.00      1.00     27035

    accuracy                           1.00     47159
   macro avg       1.00      1.00      1.00     47159
weighted avg       1.00      1.00      1.00     47159

[[20076    48]
 [   31 27004]]


In [19]:
importances = pd.Series(rf_model_v3.feature_importances_, index=feature_cols_v3)
print(importances.sort_values(ascending=False))

NoOfExternalRef      0.313923
LineOfCode           0.276963
HasSocialNet         0.134756
has_https            0.086303
num_digits           0.059665
url_length           0.043091
IsResponsive         0.033289
NoOfiFrame           0.017875
num_hyphens          0.012589
num_dots             0.010127
brand_similarity     0.008770
num_special_chars    0.002634
has_ip               0.000015
dtype: float64


**A note on ~99.8% accuracy:** this number is high enough to be suspicious, so I checked it against published research using this exact dataset. The original PhiUSIIL paper reports Random Forest reaching 99.13% accuracy, and other independent studies report similarly (up to 99.9%), so this isn't a fluke — it reflects a real, documented pattern where page-content features (external links, code length) are much more predictive than URL text alone on this dataset.

## Notebook Summary (Steps 1–7)

### What I Built
A machine learning model that classifies URLs as phishing or legitimate, using the 
PhiUSIIL dataset (235,795 URLs). I combined URL-text features I engineered myself with 
page-content features from the dataset.

### Process
1. **Explored the data** — found legitimate URLs are short and consistent (avg 26 chars), 
   phishing URLs are longer and wildly inconsistent (avg 46 chars, some over 6,000). 
   HTTPS usage was a near-perfect signal: 100% legitimate vs 49% phishing.
2. **Built 7 custom features from raw URL text** — length, digit count, hyphen count, 
   special characters, HTTPS, IP address detection. Trained and compared 3 models:

| Model | Accuracy | Phishing Recall | Missed Phishing |
|---|---|---|---|
| Logistic Regression | 94% | 0.89 | 2,182 |
| Random Forest | 96% | 0.93 | 1,392 |
| XGBoost | 96% | 0.93 | 1,409 |

3. **Error analysis** — found my model's weakness: it missed phishing sites hosted on 
   legitimate platforms (.web.app, .vercel.app) that impersonate real brands through 
   misspelled names (e.g. "amzanao" for Amazon), since these bypass my HTTPS and 
   digit-count signals.
4. **Built a brand-similarity feature** to fix this — measures how closely a domain 
   resembles known brand names. Result: missed phishing dropped from 1,392 → 941 
   (33% improvement), and I confirmed specific previously-missed URLs were now caught.
5. **Added page-content features** (external links, lines of code, social presence) — 
   pushed accuracy to ~99.8%. Verified this aligns with published research on this same 
   dataset (the original PhiUSIIL paper reports 99.13% accuracy with Random Forest), 
   so it reflects a real, documented pattern rather than an error in my process.

### Key Takeaway
Phishing pages are typically thrown-together, minimal clones — few external links, 
little code, no social presence — which content-based features capture far better than 
URL text alone. But sophisticated attacks that abuse trusted hosting platforms and 
brand-name typosquatting require more targeted features (like brand similarity) to catch.

## Choosing the Final Feature Set: Dropping Problematic Signals

After deploying this model in a live web app (`app.py`) and testing on real, popular sites, I found `NoOfExternalRef`, `LineOfCode`, and `HasSocialNet` caused false positives on famous but deliberately minimalist homepages. Google's homepage, for example, has just 41 lines of HTML and 8 external links when rendered — nowhere near my training average of ~1,947 lines / 85 links for legitimate sites. I dropped those 3 features and kept the lower-risk `IsResponsive` and `NoOfiFrame`, landing on a 10-feature model as a safer middle ground.

In [20]:
feature_cols_v4 = ['url_length', 'num_dots', 'num_hyphens', 'num_digits', 
                    'num_special_chars', 'has_https', 'has_ip', 'brand_similarity',
                    'IsResponsive', 'NoOfiFrame']

X4 = df[feature_cols_v4]
y4 = df['label']

X_train4, X_test4, y_train4, y_test4 = train_test_split(X4, y4, test_size=0.2, random_state=42)

rf_model_v4 = RandomForestClassifier(n_estimators=100, random_state=42)
rf_model_v4.fit(X_train4, y_train4)

y_pred_rf_v4 = rf_model_v4.predict(X_test4)
print(classification_report(y_test4, y_pred_rf_v4))
print(confusion_matrix(y_test4, y_pred_rf_v4))

              precision    recall  f1-score   support

           0       0.98      0.96      0.97     20124
           1       0.97      0.99      0.98     27035

    accuracy                           0.98     47159
   macro avg       0.98      0.97      0.98     47159
weighted avg       0.98      0.98      0.98     47159

[[19368   756]
 [  345 26690]]


In [21]:
importances_v4 = pd.Series(rf_model_v4.feature_importances_, index=feature_cols_v4)
print(importances_v4.sort_values(ascending=False))

has_https            0.310052
num_digits           0.166863
url_length           0.166774
NoOfiFrame           0.108768
IsResponsive         0.102377
brand_similarity     0.051760
num_dots             0.049034
num_hyphens          0.040085
num_special_chars    0.004278
has_ip               0.000009
dtype: float64


## A Deeper Bias: Live Testing Still Flagged Real Sites

Even the 10-feature model still misclassified google.com, amazon.com, github.com, discord.com, and store.steampowered.com as phishing. Tracing the raw feature values for google.com showed why: `brand_similarity` = 1.0 (a perfect match against "google" in my brand list) combined with `IsResponsive` = 0 and `NoOfiFrame` = 0 (both read as "thin," since Google's minimal page has neither) tipped the prediction toward phishing.

**Root cause:** my training data never included the real brand domains themselves (google.com, amazon.com, paypal.com, etc.) labeled as legitimate — so the model only ever learned "high brand similarity = impersonation," never "...unless it's genuinely that brand's own site." This is a real, well-known class of ML problem: a gap between what the training data represents and what the model sees in the real world.

**Plan:** patch the training data with real, popular legitimate domains — pulled from the [Tranco list](https://tranco-list.eu/), a research-grade, citable top-sites ranking — run them through the exact same feature-extraction logic the live app uses, and retrain on all 13 features.

In [22]:
from tranco import Tranco

t = Tranco(cache=True, cache_dir='.tranco')
latest_list = t.list()
top_domains = latest_list.top(1000)

print(f"Got {len(top_domains)} domains")
print(top_domains[:20])

Got 1000 domains
['google.com', 'cloudflare.com', 'gstatic.com', 'facebook.com', 'microsoft.com', 'googleapis.com', 'amazonaws.com', 'youtube.com', 'apple.com', 'gtld-servers.net', 'akamai.net', 'instagram.com', 'mail.ru', 'ezviz7.com', 'fbcdn.net', 'twitter.com', 'dzen.ru', 'linkedin.com', 'domaincontrol.com', 'googlevideo.com']


In [23]:
# Filter out infrastructure/CDN/API domains that aren't real user-facing websites.
# Doesn't need to be perfect - the fetch step below skips anything that returns
# near-empty content anyway, which catches most stragglers this list misses.
infra_keywords = ['gtld-servers', 'gstatic', 'googleapis', 'amazonaws', 'akamai', 
                   'fbcdn', 'googletagmanager', 'googlevideo', 'cloudflare-dns',
                   'edgekey', 'edgesuite', 'cloudfront', 'fastly', 'azureedge',
                   'akadns', 'domaincontrol', 'nsone', 'dnsmadeeasy']

clean_domains = [d for d in top_domains if not any(kw in d for kw in infra_keywords)]

print(f"Before filtering: {len(top_domains)}")
print(f"After filtering: {len(clean_domains)}")
print(clean_domains[:20])

Before filtering: 1000
After filtering: 978
['google.com', 'cloudflare.com', 'facebook.com', 'microsoft.com', 'youtube.com', 'apple.com', 'instagram.com', 'mail.ru', 'ezviz7.com', 'twitter.com', 'dzen.ru', 'linkedin.com', 'office.com', 'hicloudcam.com', 'live.com', 'amazon.com', 'azure.com', 'bing.com', 'github.com', 'wikipedia.org']


## Building Reusable Feature-Extraction Functions

**This is the piece that was missing before** — `extract_url_features` and `extract_page_features` were written directly into `app.py` (the web app backend) but never actually defined here in the notebook, which is why the fetch loop below couldn't find them. Defining them here, matching `app.py` exactly, means the new training examples are collected the same way the live app collects them — closing the training/serving mismatch that caused Google and friends to be misclassified.

Note: this uses simple `requests` (not the slower headless browser from an earlier version of `app.py`) so fetching ~980 domains stays practical time-wise. Some JavaScript-heavy sites may still show "thin" content here — a known limitation — but this matches what the simplified `app.py` actually uses in production, so it's the right thing to match against.

In [24]:
import re
import requests
from bs4 import BeautifulSoup

def extract_url_features(url):
    domain = url.replace('https://', '').replace('http://', '').split('/')[0]
    return {
        'url_length': len(url),
        'num_dots': url.count('.'),
        'num_hyphens': url.count('-'),
        'num_digits': sum(c.isdigit() for c in url),
        'num_special_chars': len(re.findall(r'[@%&=?]', url)),
        'has_https': 1 if url.startswith('https://') else 0,
        'has_ip': 1 if re.search(r'\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}', url) else 0,
        'brand_similarity': brand_similarity(domain)
    }

def extract_page_features(url):
    try:
        headers = {'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64)'}
        response = requests.get(url, headers=headers, timeout=5)
        html = response.text
        soup = BeautifulSoup(html, 'html.parser')
        domain = url.replace('https://', '').replace('http://', '').split('/')[0]
        links = soup.find_all('a', href=True)
        external_links = [l for l in links if l['href'].startswith('http') and domain not in l['href']]

        return {
            'NoOfExternalRef': len(external_links),
            'LineOfCode': len(html.splitlines()),
            'HasSocialNet': 1 if any(s in html.lower() for s in
                                      ['facebook.com', 'twitter.com', 'instagram.com', 'linkedin.com']) else 0,
            'IsResponsive': 1 if 'viewport' in html.lower() else 0,
            'NoOfiFrame': len(soup.find_all('iframe'))
        }
    except Exception as e:
        return {'NoOfExternalRef': 0, 'LineOfCode': 0, 'HasSocialNet': 0, 'IsResponsive': 0, 'NoOfiFrame': 0}

# Quick sanity check before running on all ~980 domains
test_url = 'https://www.google.com'
print(extract_url_features(test_url))
print(extract_page_features(test_url))

{'url_length': 22, 'num_dots': 2, 'num_hyphens': 0, 'num_digits': 0, 'num_special_chars': 0, 'has_https': 1, 'has_ip': 0, 'brand_similarity': 1.0}
{'NoOfExternalRef': 1, 'LineOfCode': 156, 'HasSocialNet': 0, 'IsResponsive': 0, 'NoOfiFrame': 0}


In [25]:
legit_feature_rows = []
failed_domains = []

for i, domain in enumerate(clean_domains):
    url = 'https://' + domain
    try:
        url_feats = extract_url_features(url)
        page_feats = extract_page_features(url)
        row = {**url_feats, **page_feats, 'label': 1}
        # Basic sanity check - skip if page barely returned anything
        if row['LineOfCode'] < 5:
            failed_domains.append(domain)
            continue
        legit_feature_rows.append(row)
    except Exception as e:
        failed_domains.append(domain)

    if i % 50 == 0:
        print(f"Processed {i}/{len(clean_domains)}... ({len(legit_feature_rows)} successful so far)")

print(f"\nDone! {len(legit_feature_rows)} successful, {len(failed_domains)} failed/skipped")

Processed 0/978... (1 successful so far)


C:\Users\AntoP\AppData\Local\Temp\ipykernel_17180\3136531341.py:23: XMLParsedAsHTMLWarning: It looks like you're using an HTML parser to parse an XML document.

Assuming this really is an XML document, what you're doing might work, but you should know that using an XML parser will be more reliable. To parse this document as XML, make sure you have the Python package 'lxml' installed, and pass the keyword argument `features="xml"` into the BeautifulSoup constructor.

If you want or need to use an HTML parser on this document, you can make this warning go away by filtering it. To do that, run this code before calling the BeautifulSoup constructor:

    from bs4 import XMLParsedAsHTMLWarning
    import warnings

    warnings.filterwarnings("ignore", category=XMLParsedAsHTMLWarning)

  soup = BeautifulSoup(html, 'html.parser')


Processed 50/978... (35 successful so far)
Processed 100/978... (69 successful so far)
Processed 150/978... (98 successful so far)
Processed 200/978... (126 successful so far)
Processed 300/978... (184 successful so far)
Processed 350/978... (216 successful so far)
Processed 400/978... (241 successful so far)
Processed 500/978... (289 successful so far)
Processed 750/978... (437 successful so far)
Processed 800/978... (469 successful so far)
Processed 900/978... (523 successful so far)

Done! 570 successful, 408 failed/skipped


## Merging the Patch Data and Retraining

Since ~900+ new rows is tiny next to the 235,795-row original dataset, each one gets duplicated several times so the model actually weighs them meaningfully instead of treating them as noise.

In [26]:
# Turn the fetched rows into a DataFrame matching the training columns
patch_df = pd.DataFrame(legit_feature_rows)
print(f"Patch dataset: {len(patch_df)} verified real legitimate domains")

# Duplicate each row 8x so these examples carry real weight against 235,795 original rows
patch_df_boosted = pd.concat([patch_df] * 8, ignore_index=True)
print(f"After duplication: {len(patch_df_boosted)} rows")

# Combine with the original training features (13-feature set) + label
original_subset = df[feature_cols_v3 + ['label']].copy()
combined_df = pd.concat([original_subset, patch_df_boosted[feature_cols_v3 + ['label']]], ignore_index=True)
print(f"Combined training data: {len(combined_df)} rows")

Patch dataset: 570 verified real legitimate domains
After duplication: 4560 rows
Combined training data: 240355 rows


In [27]:
X5 = combined_df[feature_cols_v3]
y5 = combined_df['label']

X_train5, X_test5, y_train5, y_test5 = train_test_split(X5, y5, test_size=0.2, random_state=42)

rf_model_v5 = RandomForestClassifier(n_estimators=100, random_state=42)
rf_model_v5.fit(X_train5, y_train5)

y_pred_rf_v5 = rf_model_v5.predict(X_test5)
print(classification_report(y_test5, y_pred_rf_v5))
print(confusion_matrix(y_test5, y_pred_rf_v5))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00     20039
           1       1.00      1.00      1.00     28032

    accuracy                           1.00     48071
   macro avg       1.00      1.00      1.00     48071
weighted avg       1.00      1.00      1.00     48071

[[19979    60]
 [   36 27996]]


## Re-Testing the Original Tricky URLs

In [31]:
tricky_urls = ['https://www.google.com', 'https://www.amazon.com', 'https://github.com',
               'https://discord.com', 'https://store.steampowered.com', 'https://www.paypal.com',
               'https://www.microsoft.com', 'https://www.nasa.gov']

for url in tricky_urls:
    url_feats = extract_url_features(url)
    page_feats = extract_page_features(url)
    all_feats = {**url_feats, **page_feats}
    feature_values = [[all_feats[f] for f in feature_cols_v3]]
    pred = rf_model_v5.predict(feature_values)[0]
    proba = rf_model_v5.predict_proba(feature_values)[0]
    result = 'Legitimate' if pred == 1 else 'Phishing'
    print(url, '->', result, f'({round(max(proba)*100,1)}% confidence)')

C:\Users\AntoP\AppData\Local\Python\pythoncore-3.14-64\Lib\site-packages\sklearn\utils\validation.py:2827: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(
C:\Users\AntoP\AppData\Local\Python\pythoncore-3.14-64\Lib\site-packages\sklearn\utils\validation.py:2827: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(
C:\Users\AntoP\AppData\Local\Python\pythoncore-3.14-64\Lib\site-packages\sklearn\utils\validation.py:2827: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(
C:\Users\AntoP\AppData\Local\Python\pythoncore-3.14-64\Lib\site-packages\sklearn\utils\validation.py:2827: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(


https://www.google.com -> Legitimate (95.0% confidence)
https://www.amazon.com -> Phishing (99.0% confidence)


C:\Users\AntoP\AppData\Local\Python\pythoncore-3.14-64\Lib\site-packages\sklearn\utils\validation.py:2827: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(
C:\Users\AntoP\AppData\Local\Python\pythoncore-3.14-64\Lib\site-packages\sklearn\utils\validation.py:2827: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(
C:\Users\AntoP\AppData\Local\Python\pythoncore-3.14-64\Lib\site-packages\sklearn\utils\validation.py:2827: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(
C:\Users\AntoP\AppData\Local\Python\pythoncore-3.14-64\Lib\site-packages\sklearn\utils\validation.py:2827: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(


https://github.com -> Legitimate (100.0% confidence)
https://discord.com -> Legitimate (100.0% confidence)


C:\Users\AntoP\AppData\Local\Python\pythoncore-3.14-64\Lib\site-packages\sklearn\utils\validation.py:2827: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(
C:\Users\AntoP\AppData\Local\Python\pythoncore-3.14-64\Lib\site-packages\sklearn\utils\validation.py:2827: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(


https://store.steampowered.com -> Legitimate (100.0% confidence)


C:\Users\AntoP\AppData\Local\Python\pythoncore-3.14-64\Lib\site-packages\sklearn\utils\validation.py:2827: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(
C:\Users\AntoP\AppData\Local\Python\pythoncore-3.14-64\Lib\site-packages\sklearn\utils\validation.py:2827: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(


https://www.paypal.com -> Legitimate (97.0% confidence)


C:\Users\AntoP\AppData\Local\Python\pythoncore-3.14-64\Lib\site-packages\sklearn\utils\validation.py:2827: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(
C:\Users\AntoP\AppData\Local\Python\pythoncore-3.14-64\Lib\site-packages\sklearn\utils\validation.py:2827: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(


https://www.microsoft.com -> Legitimate (100.0% confidence)
https://www.nasa.gov -> Legitimate (100.0% confidence)


C:\Users\AntoP\AppData\Local\Python\pythoncore-3.14-64\Lib\site-packages\sklearn\utils\validation.py:2827: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(
C:\Users\AntoP\AppData\Local\Python\pythoncore-3.14-64\Lib\site-packages\sklearn\utils\validation.py:2827: UserWarning: X does not have valid feature names, but RandomForestClassifier was fitted with feature names
  warnings.warn(


If these all now correctly show **Legitimate**, save this as the final model:

In [29]:
import pickle

with open('phishing_model.pkl', 'wb') as f:
    pickle.dump(rf_model_v5, f)

print("Patched model saved as phishing_model.pkl")

Patched model saved as phishing_model.pkl


## Current Status & Next Steps

1. Run the cells above in order — the Tranco fetch loop takes roughly 15–30+ minutes since it's fetching ~980 real, live pages.
2. Check the "Re-Testing the Original Tricky URLs" output — confirm Google, Amazon, GitHub, Discord, Steam, PayPal, Microsoft, and NASA all show **Legitimate**.
3. If they do, `phishing_model.pkl` is now updated — restart `app.py` (it already expects a 13-feature model matching `feature_cols_v3`, but double check the feature list in `app.py` matches `extract_page_features` returning all 5 page-content features, not just `IsResponsive`/`NoOfiFrame`).
4. Re-run `test.py` with the full 15-URL list to confirm both the tricky major sites AND the original error-analysis edge cases still work well.
5. Update the final writeup with this discovery — the brand-matching training gap is genuinely one of the strongest, most sophisticated findings in this whole project.

In [30]:
   patch_df.to_csv('legit_domains_patch.csv', index=False)
   print(f"Saved {len(patch_df)} verified domains to legit_domains_patch.csv")

Saved 570 verified domains to legit_domains_patch.csv


In [32]:
print(patch_df[patch_df['label'] == 1].shape)
amazon_check = [d for d in clean_domains if 'amazon' in d]
print("Amazon-related domains in Tranco list:", amazon_check)

(570, 14)
Amazon-related domains in Tranco list: ['amazon.com', 'amazonvideo.com', 'amazon-adsystem.com', 'amazon.dev', 'media-amazon.com', 'amazon.co.uk', 'amazon.de', 'amazontrust.com', 'amazonalexa.com', 'amazon.co.jp', 'amazon.co.za', 'amazon.fr', 'amazon.ca', 'amazon.in', 'amazon.it', 'amazon.es', 'amazon.com.au', 'amazon.com.br', 'ssl-images-amazon.com']


In [33]:
amazon_in_patch = patch_df[patch_df['Domain'].str.contains('amazon.com', case=False, na=False)] if 'Domain' in patch_df.columns else "no Domain column"
print(amazon_in_patch)

# If there's no Domain column, check differently - let's see what columns patch_df actually has
print(patch_df.columns.tolist())

no Domain column
['url_length', 'num_dots', 'num_hyphens', 'num_digits', 'num_special_chars', 'has_https', 'has_ip', 'brand_similarity', 'NoOfExternalRef', 'LineOfCode', 'HasSocialNet', 'IsResponsive', 'NoOfiFrame', 'label']


In [34]:
high_brand_match = patch_df[patch_df['brand_similarity'] > 0.9]
print(f"Rows with near-perfect brand match: {len(high_brand_match)}")
print(high_brand_match)

Rows with near-perfect brand match: 15
     url_length  num_dots  num_hyphens  num_digits  num_special_chars  \
0            18         1            0           0                  0   
2            20         1            0           0                  0   
3            21         1            0           0                  0   
5            17         1            0           0                  0   
6            21         1            0           0                  0   
9            20         1            0           0                  0   
18           19         1            0           0                  0   
85           19         1            0           0                  0   
94           18         1            0           0                  0   
208          17         1            0           0                  0   
212          19         1            0           0                  0   
281          17         1            0           0                  0   
285         

In [35]:
amazon_url_feats = extract_url_features('https://www.amazon.com')
amazon_page_feats = extract_page_features('https://www.amazon.com')
print("Amazon's current live features:", {**amazon_url_feats, **amazon_page_feats})

Amazon's current live features: {'url_length': 22, 'num_dots': 2, 'num_hyphens': 0, 'num_digits': 0, 'num_special_chars': 0, 'has_https': 1, 'has_ip': 0, 'brand_similarity': 1.0, 'NoOfExternalRef': 0, 'LineOfCode': 0, 'HasSocialNet': 0, 'IsResponsive': 0, 'NoOfiFrame': 0}


In [37]:
def extract_page_features(url):
    try:
        headers = {'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64)'}
        response = requests.get(url, headers=headers, timeout=5)
        
        # NEW: check if the request actually succeeded, not just "didn't crash"
        if response.status_code != 200:
            print(f"  Fetch failed for {url}: status code {response.status_code}")
            return None
        
        html = response.text
        soup = BeautifulSoup(html, 'html.parser')
        domain = url.replace('https://', '').replace('http://', '').split('/')[0]
        links = soup.find_all('a', href=True)
        external_links = [l for l in links if l['href'].startswith('http') and domain not in l['href']]

        return {
            'NoOfExternalRef': len(external_links),
            'LineOfCode': len(html.splitlines()),
            'HasSocialNet': 1 if any(s in html.lower() for s in
                                      ['facebook.com', 'twitter.com', 'instagram.com', 'linkedin.com']) else 0,
            'IsResponsive': 1 if 'viewport' in html.lower() else 0,
            'NoOfiFrame': len(soup.find_all('iframe'))
        }
    except Exception as e:
        print(f"  Fetch failed for {url}: {type(e).__name__} - {e}")
        return None

In [38]:
legit_feature_rows = []
failed_domains = []

for i, domain in enumerate(clean_domains):
    url = 'https://' + domain
    try:
        url_feats = extract_url_features(url)
        page_feats = extract_page_features(url)
        
        # NEW: skip this domain entirely if the page fetch failed
        if page_feats is None:
            failed_domains.append(domain)
            continue
        
        row = {**url_feats, **page_feats, 'label': 1}
        if row['LineOfCode'] < 5:
            failed_domains.append(domain)
            continue
        legit_feature_rows.append(row)
    except Exception as e:
        failed_domains.append(domain)

    if i % 50 == 0:
        print(f"Processed {i}/{len(clean_domains)}... ({len(legit_feature_rows)} successful so far)")

print(f"\nDone! {len(legit_feature_rows)} successful, {len(failed_domains)} failed/skipped")

Processed 0/978... (1 successful so far)
  Fetch failed for https://ezviz7.com: ConnectionError - HTTPSConnectionPool(host='ezviz7.com', port=443): Max retries exceeded with url: / (Caused by NameResolutionError("HTTPSConnection(host='ezviz7.com', port=443): Failed to resolve 'ezviz7.com' ([Errno 11001] getaddrinfo failed)"))
  Fetch failed for https://hicloudcam.com: ConnectionError - HTTPSConnectionPool(host='hicloudcam.com', port=443): Max retries exceeded with url: / (Caused by NameResolutionError("HTTPSConnection(host='hicloudcam.com', port=443): Failed to resolve 'hicloudcam.com' ([Errno 11001] getaddrinfo failed)"))
  Fetch failed for https://apple-dns.net: ConnectionError - HTTPSConnectionPool(host='apple-dns.net', port=443): Max retries exceeded with url: / (Caused by NameResolutionError("HTTPSConnection(host='apple-dns.net', port=443): Failed to resolve 'apple-dns.net' ([Errno 11001] getaddrinfo failed)"))
  Fetch failed for https://googleusercontent.com: status code 404
  Fe

In [39]:
patch_df = pd.DataFrame(legit_feature_rows)
patch_df.to_csv('legit_domains_patch.csv', index=False)
print(f"Saved {len(patch_df)} verified domains to legit_domains_patch.csv")

Saved 523 verified domains to legit_domains_patch.csv


In [41]:
import numpy as np

original_subset = df[feature_cols_v3 + ['label']].copy()
patch_subset = patch_df[feature_cols_v3 + ['label']].copy()

combined_df_v2 = pd.concat([original_subset, patch_subset], ignore_index=True)

weights = np.ones(len(combined_df_v2))
weights[len(original_subset):] = 100

X6 = combined_df_v2[feature_cols_v3]
y6 = combined_df_v2['label']

X_train6, X_test6, y_train6, y_test6, w_train6, w_test6 = train_test_split(
    X6, y6, weights, test_size=0.2, random_state=42
)

rf_model_v6 = RandomForestClassifier(n_estimators=100, random_state=42)
rf_model_v6.fit(X_train6, y_train6, sample_weight=w_train6)

y_pred_rf_v6 = rf_model_v6.predict(X_test6)
print(classification_report(y_test6, y_pred_rf_v6))
print(confusion_matrix(y_test6, y_pred_rf_v6))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00     20122
           1       1.00      1.00      1.00     27142

    accuracy                           1.00     47264
   macro avg       1.00      1.00      1.00     47264
weighted avg       1.00      1.00      1.00     47264

[[20068    54]
 [   48 27094]]


In [42]:
import warnings
warnings.filterwarnings('ignore')

for url in tricky_urls:
    url_feats = extract_url_features(url)
    page_feats = extract_page_features(url)
    
    if page_feats is None:
        print(url, '-> COULD NOT FETCH PAGE (likely blocked or timed out)')
        continue
    
    all_feats = {**url_feats, **page_feats}
    feature_values = [[all_feats[f] for f in feature_cols_v3]]
    pred = rf_model_v6.predict(feature_values)[0]
    proba = rf_model_v6.predict_proba(feature_values)[0]
    result = 'Legitimate' if pred == 1 else 'Phishing'
    print(url, '->', result, f'({round(max(proba)*100,1)}% confidence)')

https://www.google.com -> Legitimate (91.0% confidence)
  Fetch failed for https://www.amazon.com: status code 202
https://www.amazon.com -> COULD NOT FETCH PAGE (likely blocked or timed out)
https://github.com -> Legitimate (100.0% confidence)
https://discord.com -> Legitimate (100.0% confidence)
https://store.steampowered.com -> Legitimate (100.0% confidence)
https://www.paypal.com -> Legitimate (100.0% confidence)
https://www.microsoft.com -> Legitimate (100.0% confidence)
https://www.nasa.gov -> Legitimate (100.0% confidence)


In [43]:
%run train_model.py

Saved phishing_model.pkl using 235,795 original rows and 523 verified domains.
